src/map/entities/battleentity.cpp

```diff
@@ -1263,6 +1263,14 @@ void CBattleEntity::OnCastFinished(CMagicState& state, action_t& action)
         }
     }
 
+    if (this != PActionTarget && PActionTarget->objtype == TYPE_PC 
```

Comment:
just check if their allegiances are not the same, and you can skip the isbuff/isheal/isna, I think